node/full_node_test.go

```diff
@@ -73,30 +69,58 @@ func TestMempoolDirectly(t *testing.T) {
 		assert.NoError(node.Stop())
 	}()
 
-	pid, err := peer.IDFromPrivateKey(anotherKey)
-	require.NoError(err)
-	err = node.Mempool.CheckTx([]byte("tx1"), func(r *abci.Response) {}, mempool.TxInfo{
-		SenderID: node.mempoolIDs.GetForPeer(pid),
-	})
-	require.NoError(err)
-	err = node.Mempool.CheckTx([]byte("tx2"), func(r *abci.Response) {}, mempool.TxInfo{
-		SenderID: node.mempoolIDs.GetForPeer(pid),
-	})
-	require.NoError(err)
-	time.Sleep(100 * time.Millisecond)
-	err = node.Mempool.CheckTx([]byte("tx3"), func(r *abci.Response) {}, mempool.TxInfo{
-		SenderID: node.mempoolIDs.GetForPeer(pid),
-	})
-	require.NoError(err)
-	err = node.Mempool.CheckTx([]byte("tx4"), func(r *abci.Response) {}, mempool.TxInfo{
-		SenderID: node.mempoolIDs.GetForPeer(pid),
-	})
+	pid := getPeerID(anotherKey, require)
+
+	checkTransactions(node, pid, require)
+
+	checkMempoolSize(node, require)
+}
+
+func setupMockApplication() *mocks.Application {
+	app := &mocks.Application{}
+	app.On(InitChain, mock.Anything).Return(abci.ResponseInitChain{})
+	app.On(CheckTx, mock.Anything).Return(abci.ResponseCheckTx{})
+	return app
+}
+
+func generateKeys() (crypto.PrivKey, crypto.PrivKey) {
+	key, _, _ := crypto.GenerateEd25519Key(rand.Reader)
+	signingKey, _, _ := crypto.GenerateEd25519Key(rand.Reader)
+	return key, signingKey
+}
+
+func generateSingleKey() crypto.PrivKey {
+	key, _, _ := crypto.GenerateEd25519Key(rand.Reader)
+	return key
+}
+
+func setupNode(ctx context.Context, key crypto.PrivKey, signingKey crypto.PrivKey, app *mocks.Application) (*FullNode, error) {
```

Comment:
nit: I'd name this `newTestFullNode` since you aren't doing any set up, you are just call `newFullNode`. 
`setupNode` to me would imply we are doing some additional set up like initializing the app, connecting to peers, waiting for x number of blocks, etc.

Reply:
so you might do the following
```go
func setupTestNode (ctx) {
	app := setupMockApplication()
	key, signingKey := generateKeys()
	anotherKey := generateSingleKey()
        node, err := newTestFullNode()

       return node, err
}
```

since you only need the node and the context in the test. 

then the test looks like (ignoring the obvious syntax errors of my shorthand)

```go

func TestXYZ() {
  ctx := context.Background()
  node, err := setupTestNode(ctx)
  defer ctx.cancel
  defer node.stop()

   /// test code
}
```

Reply:
Thank you, all are great suggestions. Applied them.